Sources/Bot/Services/DispatchService.swift

```diff
@@ -0,0 +1,198 @@
+        topPriorityLabels: [PullRequest.Label],
+        requiresAllStatusChecks: Bool,
+        statusChecksTimeout: TimeInterval,
+        logger: LoggerProtocol,
+        gitHubAPI: GitHubAPIProtocol,
+        gitHubEvents: GitHubEventsServiceProtocol,
+        scheduler: DateScheduler = QueueScheduler(),
+        mergeServiceLifecycleObserver: Signal<MergeServiceLifecycleEvent, NoError>.Observer? = nil
+    ) {
+        self.integrationLabel = integrationLabel
+        self.topPriorityLabels = topPriorityLabels
+        self.requiresAllStatusChecks = requiresAllStatusChecks
+        self.statusChecksTimeout = statusChecksTimeout
+
+        self.logger = logger
+        self.gitHubAPI = gitHubAPI
+        self.scheduler = scheduler
+
+        self.mergeServices = [:]
+        self.mergeServiceLifecycleObserver = mergeServiceLifecycleObserver
+
+        healthcheck = Healthcheck(scheduler: scheduler)
+
+        gitHubAPI.fetchPullRequests()
+            .flatMapError { _ in .value([]) }
+            .map { pullRequests in
+                pullRequests.filter { $0.isLabelled(with: self.integrationLabel) }
+            }
+            .observe(on: scheduler)
+            .startWithValues { pullRequests in
+                self.dispatchInitial(pullRequests: pullRequests)
+            }
+
+        gitHubEvents.events
+            .observe(on: scheduler)
+            .observeValues { [weak self] gitHubEvent in
+                switch gitHubEvent {
+                case let .pullRequest(event):
+                    self?.pullRequestDidChange(event: event)
+                case let .status(event):
+                    self?.statusChecksDidChange(event: event)
+                case .ping:
+                    break
+                }
+            }
+    }
+
+    private func dispatchInitial(pullRequests: [PullRequest]) {
+        let dispatchTable = Dictionary(grouping: pullRequests) { pullRequest in pullRequest.target.ref }
+        for (branch, pullRequestsForBranch) in dispatchTable {
+            makeMergeService(targetBranch: branch, scheduler: self.scheduler, initialPullRequests: pullRequestsForBranch)
+        }
+    }
+
+    private func pullRequestDidChange(event: PullRequestEvent) {
+        logger.log("📣 Pull Request did change \(event.pullRequestMetadata) with action `\(event.action)`")
+        let targetBranch = event.pullRequestMetadata.reference.target.ref
+        if let service = mergeServices[targetBranch] {
+            service.pullRequestChangesObserver.send(value: (event.pullRequestMetadata, event.action))
+        } else {
+            makeMergeService(targetBranch: targetBranch, scheduler: self.scheduler, initialPullRequests: [event.pullRequestMetadata.reference])
+        }
+    }
+
+    private func statusChecksDidChange(event: StatusEvent) {
+        // No way to know which MergeService this event is supposed to be for – isRelative(toBranch:) only checks for head branch not target so not useful here
```

Comment:
so there is no target branch in the PR metadata with the event? 🤔

Reply:
If I understand [the doc](https://developer.github.com/v3/activity/events/types/#statusevent) correctly, only the list of branches this check _affects_, so the _source_ branches, are part of the payloed, but not target branches?

Reply:
not sure, as in example there are multiple branches. Can we validate that? Maybe using our fork of the repo?

Reply:
Well on line 486 of `MergeService.swift` we do:

```swift
            return statusChecksCompletion
                .observe(on: scheduler)
                .filter { change in change.state != .pending && change.isRelative(toBranch: pullRequest.source.ref) }
                .on { change in
                    logger.log("📣 Status check `\(change.context)` finished with result: `\(change.state)` (SHA: `\(change.sha)`)")
                }
```
Which means we use that `isRelative(to:)` method to check against… the `source` branch (not the target branch), which should confirm my understanding at least…

And since the `MergeService` do filter those status checks (which it can do because it has the reference to all its PRs which reference said source branch so can use that to filter, as opposed to `DispatchService` which only know the `targetBranch` of each `MergeService`), not sure that really matters anyway 😉